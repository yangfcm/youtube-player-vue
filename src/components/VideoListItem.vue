<script setup lang="ts">
import { useDisplay } from 'vuetify';
import type { PlayListItemDetails } from '@/stores/types';

defineProps<{
  video: PlayListItemDetails
}>()

const { name } = useDisplay();
</script>

<template>
  <v-card class="mb-3">
    <div class="d-flex flex-column flex-sm-row">
      <v-avatar 
        rounded="0"
        :style="`height: auto; width: ${name === 'xs' ? '100%' : '230px'}`"
      >
        <v-img :src="video.snippet.thumbnails.high?.url"></v-img>
      </v-avatar>
      <div>
        <v-card-title :title="video.snippet.title">{{ video.snippet.title }}</v-card-title>
      </div>
    </div>
  </v-card>
</template>