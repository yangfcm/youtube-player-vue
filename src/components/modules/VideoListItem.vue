<template>
  <div class="ui card app-list-card">
    <router-link
      class="app-card-image-container"
      :to="
        video.snippet.playlistId
          ? `/video/${videoId}?playlistId=${video.snippet.playlistId}`
          : `/video/${videoId}`
      "
    >
      <img
        :src="video.snippet.thumbnails.medium.url"
        :alt="video.snippet.title"
        class="app-list-card-image"
      />
    </router-link>
    <div class="content app-list-card-content">
      <router-link
        :to="
          video.snippet.playlistId
            ? `/video/${videoId}?playlistId=${video.snippet.playlistId}`
            : `/video/${videoId}`
        "
        class="header app-list-card-title"
      >{{ video.snippet.title }}</router-link>
      <router-link :to="`/channel/${video.snippet.channelId}`">
        {{
        video.snippet.channelTitle
        }}
      </router-link>
      <div class="meta">
        <i class="calendar icon"></i>
        {{ video.snippet.publishedAt | date }}
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: ["video"],
  computed: {
    videoId() {
      let videoId = "";
      if (this.video && this.video.snippet.resourceId) {
        videoId = this.video.snippet.resourceId.videoId;
      } else if (this.video && this.video.id) {
        videoId = this.video.id.videoId;
      }
      return videoId;
    },
  },
};
</script>

<style scoped>
.app-list-card {
  display: flex;
  flex-direction: row;
  width: 100%;
  margin-bottom: 10px !important;
  border: none;
  box-shadow: none;
}
.app-card-image-container {
  display: flex;
  align-items: center;
}
.app-list-card-image {
  max-width: 140px;
  height: auto;
}
.app-list-card-title {
  line-height: 23px;
  max-height: 46px;
  overflow: hidden;
  text-overflow: ellipsis;
}
.app-list-card-content {
  flex-grow: 1;
  border: none !important;
}
</style>
