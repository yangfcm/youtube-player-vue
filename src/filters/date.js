/**
 * Convert a date object to human-readable string
 */
export default (date) => {
  const dateObj = new Date(date);
  const day = dateObj.getDate();
  const month = new Intl.DateTimeFormat("en-US", {
    month: "short",
  }).format(dateObj);
  const year = dateObj.getFullYear();
  return `${day} ${month}, ${year}`;
};
