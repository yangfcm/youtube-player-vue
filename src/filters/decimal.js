/**
 * Separate an integer number by commas
 * e.g. 12345678 => 12,345,678
 */
export default (num) => {
  if (isNaN(Number(num))) return "";
  let strNum = String(num);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
