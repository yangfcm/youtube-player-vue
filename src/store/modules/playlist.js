import axios from "../../settings";
import {
  FETCH_PLAY_LIST,
  FETCH_PLAY_LIST_DETAIL,
  CATCH_ERROR,
  FETCH_MY_PLAY_LIST,
} from "../types";
import parseError from "../helpers/parseError";

const state = {
  playlist: null,
  myPlaylist: null,
  playlistDetail: null,
  playlistError: null,
};

const getters = {
  playlist: (state) => state.playlist,
  myPlaylist: (state) => state.myPlaylist,
  playlistDetail: (state) => state.playlistDetail,
  playlistError: (state) =>
    parseError(state.playlistError, "Failed to fetch playlist videos"),
};

const mutations = {
  FETCH_PLAY_LIST: (state, payload) => {
    state.playlist = payload;
    state.playlistError = null;
  },
  FETCH_MY_PLAY_LIST: (state, payload) => {
    state.myPlaylist = payload;
    state.playlistError = null;
  },
  FETCH_PLAY_LIST_DETAIL: (state, payload) => {
    state.playlistDetail = payload;
    state.playlistError = null;
  },
  CATCH_ERROR: (state, payload) => {
    state.playlistError = payload;
  },
};

const actions = {
  /** Fetch the playlist of authed user */
  fetchMyPlaylist: async (context, pageToken = null) => {
    const accessToken = localStorage.getItem("access_token");
    try {
      const response = await axios.get("/playlists", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet,contentDetails",
          maxResults: 12,
          pageToken,
          mine: true,
        },
      });
      context.commit(FETCH_MY_PLAY_LIST, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },

  /** Fetch the playlist that belongs to a channel */
  fetchPlaylist: async (context, channelId, pageToken = null) => {
    try {
      const response = await axios.get("/playlists", {
        params: {
          ...axios.defaults.params,
          part: "snippet,contentDetails",
          maxResults: 12,
          channelId,
          pageToken,
        },
      });
      context.commit(FETCH_PLAY_LIST, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },

  /** Fetch the videos in a playlist */
  fetchPlaylistDetail: async (context, playlistId, pageToken = null) => {
    try {
      const response = await axios.get("/playlistItems", {
        params: {
          ...axios.defaults.params,
          part: "snippet,contentDetails",
          maxResults: 10,
          playlistId,
          pageToken,
        },
      });
      context.commit(FETCH_PLAY_LIST_DETAIL, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
