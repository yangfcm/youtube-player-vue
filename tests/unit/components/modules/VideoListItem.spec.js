import { shallowMount, RouterLinkStub, createLocalVue } from "@vue/test-utils";
import VideoListItem from "@/components/modules/VideoListItem";
import { playlistDetailResponse } from "../../store/fixtures/playlist";
import date from "@/filters/date";
import decimal from "@/filters/decimal";

describe("Test VideoListItem component", () => {
  const localVue = createLocalVue();
  localVue.filter("date", date);
  localVue.filter("decimal", decimal);
  let wrapper;
  const video = playlistDetailResponse.items[0];
  beforeEach(() => {
    wrapper = shallowMount(VideoListItem, {
      propsData: {
        video,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      localVue,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("img").attributes("src")).toBe(
      video.snippet.thumbnails.medium.url
    );
    expect(wrapper.find("a.app-card-image-container").props().to).toBe(
      `/video/${video.snippet.resourceId.videoId}?playlistId=${video.snippet.playlistId}`
    );
    expect(wrapper.find("div.app-list-card-content a").props().to).toBe(
      `/video/${video.snippet.resourceId.videoId}?playlistId=${video.snippet.playlistId}`
    );
    expect(wrapper.find("div.app-list-card-content").text()).toContain(
      video.snippet.title
    );
    expect(wrapper.find("a.app-channel-link").props().to).toBe(
      `/channel/${video.snippet.channelId}`
    );
    expect(wrapper.find("a.app-channel-link").text()).toContain(
      video.snippet.channelTitle
    );
  });
});
