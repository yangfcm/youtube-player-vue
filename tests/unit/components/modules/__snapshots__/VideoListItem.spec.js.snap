// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoListItem component should match snapshot 1`] = `
<div class="ui card app-list-card"><a class="app-card-image-container"><img src="https://i.ytimg.com/vi/SqrbIlUwR0U/mqdefault.jpg" alt="Go / Golang Crash Course" class="app-list-card-image"></a>
  <div class="content app-list-card-content"><a class="header app-list-card-title">Go / Golang Crash Course</a> <a class="app-channel-link">
      Fan Yang
    </a>
    <div class="meta"><i class="calendar icon"></i>
      21 Sep, 2018
    </div>
  </div>
</div>
`;
